PR: https://github.com/odoo/odoo/pull/239681

From: 18114903afc3ebf011ebb3b55e4a0d542162a3e3
From: Stephane Vanmeerhaeghe (stva)
Date: 2025-12-19 10:32:37

Breaking data model changes scores: del:9 + add:0, change matches:
-    background_image = fields.Binary('Background Image')
-    background_color = fields.Char('Background Color', help='The background color of the floor in a html-compatible format')
-    floor_background_image = fields.Image(string='Floor Background Image')
-    shape = fields.Selection([('square', 'Square'), ('round', 'Round')], string='Shape', required=True, default='square')
-    position_h = fields.Float('Horizontal Position', default=10,
-    position_v = fields.Float('Vertical Position', default=10,
-    width = fields.Float('Width', default=50, help="The table's width in pixels")
-    height = fields.Float('Height', default=50, help="The table's height in pixels")
-    color = fields.Char('Color', help="The table's color, expressed as a valid 'background' CSS property value")

Total Changes: 260

[IMP] pos_restaurant:  enhance floor plan editor

This commit adds the ability for users to create decorative elements within the floor plan editor.
It introduces shapes, text items, and images that can be placed and edited directly on the floor plan.
These enhancements significantly improve customization options, enabling users to design floor layouts more freely than before.

Task.5159581

closes odoo/odoo#239681

Related: odoo/enterprise#101941
Related: odoo/upgrade#9065
Signed-off-by: Stéphane Vanmeerhaeghe (stva) <stva@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_restaurant/models/pos_config.py
+++ b/addons/pos_restaurant/models/pos_config.py
@@ -12,6 +12,8 @@ class PosConfig(models.Model):
     floor_ids = fields.Many2many('restaurant.floor', string='Restaurant Floors', help='The restaurant floors served by this point of sale.', copy=False)
     default_screen = fields.Selection([('tables', 'Tables'), ('register', 'Register')], string='Default Screen', default='tables')
     use_course_allocation = fields.Boolean(string="Enable Course Allocation")
+    floor_plan_settings = fields.Json(string='Floor Plan Settings')
+    floor_plan = fields.Json(string='Floor Plan', compute="_compute_floor_plan")
 
     def _get_forbidden_change_fields(self):
         forbidden_keys = super()._get_forbidden_change_fields()
@@ -54,10 +56,7 @@ class PosConfig(models.Model):
                 'table_number': 1,
                 'floor_id': main_floor.id,
                 'seats': 1,
-                'position_h': 100,
-                'position_v': 100,
-                'width': 130,
-                'height': 130,
+                'floor_plan_layout': {'top': 100, 'left': 100, 'width': 130, 'height': 130, 'color': 'green', 'shape': 'square'},
             })
 
     @api.model
@@ -142,6 +141,10 @@ class PosConfig(models.Model):
             convert.convert_file(self._env_with_clean_context(), 'pos_restaurant', 'data/scenarios/restaurant_demo_session.xml', idref=None, mode='init', noupdate=True)
         return {'config_id': config.id}
 
+    @api.depends('floor_plan_settings')
+    def _compute_local_data_integrity(self):
+        super()._compute_local_data_integrity()
+
     def _load_restaurant_demo_data(self, with_demo_data=True):
         self.ensure_one()
         convert.convert_file(self._env_with_clean_context(), 'pos_restaurant', 'data/scenarios/restaurant_category_data.xml', idref=None, mode='init', noupdate=True)
@@ -167,3 +170,164 @@ class PosConfig(models.Model):
         if self.module_pos_restaurant:
             return 'pos_restaurant.pos_config_main_restaurant'
         return super()._get_default_demo_data_xml_id()
+
+    def _compute_floor_plan(self):
+        for record in self:
+            if not record.module_pos_restaurant:
+                record.floor_plan = None
+                continue
+            record.floor_plan = {
+                'settings': record.floor_plan_settings or {},
+                'floors': [
+                    {
+                        'id': floor.id,
+                        **(floor.floor_plan_layout or {}),
+                        'tables': [
+                            {'id': table.id, **(table.floor_plan_layout or {})}
+                            for table in floor.table_ids if table.active
+                        ]
+                    }
+                    for floor in record.floor_ids if floor.active
+                ],
+            }
+
+    def get_floor_plan(self):
+        self.ensure_one()
+        return {
+            'config': {
+                'floor_plan': self.floor_plan,
+                'floor_ids': self.floor_ids.ids,
+            },
+            'records': {
+                'restaurant.floor': self.env['restaurant.floor']._load_pos_data_read(self.floor_ids, self),
+                'restaurant.table': self.env['restaurant.table']._load_pos_data_read(self.floor_ids.table_ids, self),
+            }
+        }
+
+    def save_floor_plan(self, floor_plan):
+        self.ensure_one()
+        existing_floors = {f.id: f for f in self.floor_ids if f.active}
+        existing_tables = {t.id: t for t in self.floor_ids.table_ids if t.active}
+        seen_floor_ids = set()
+        seen_table_ids = set()
+        session_id = self.current_session_id.id
+
+        for floor_data in floor_plan.get("floors", []):
+            floor = self._save_floor_plan_floor(floor_data, existing_floors)
+            if not floor:
+                continue
+            seen_floor_ids.add(floor.id)
+
+            for table_data in floor_data.get("tables", []):
+                table = self._save_floor_plan_table(table_data, floor, existing_tables)
+                if table:
+                    seen_table_ids.add(table.id)
+
+        self.floor_plan_settings = floor_plan.get("settings", {})
+
+        # Deactivate removed floors and tables
+        for floor_id, floor in existing_floors.items():
+            if floor_id not in seen_floor_ids:
+                floor.deactivate_floor(session_id)
+
+        for table_id, table in existing_tables.items():
+            if table_id not in seen_table_ids:
+                table.are_orders_still_in_draft()
+                table.active = False
+
+        config_ids = self.floor_ids.pos_config_ids
+
+        for config in config_ids:
+            config._notify("FLOOR_PLAN_UPD", {
+                'device_identifier': self.env.context.get('device_identifier', False),
+                'session_id': session_id
+            })
+
+        return self.get_floor_plan()
+
+    def _save_floor_plan_floor(self, floor_data, existing_floors):
+        self.ensure_one()
+        floor_id = floor_data.get("id")
+        model_data, layout_data = self._split_layout_server_data(floor_data, {"name"}, {"id", "tables", "newImages"})
+        if floor_id:
+            floor = existing_floors.get(floor_id)
+            if not floor:
+                return None
+            floor.write({**model_data})
+        else:
+            floor = self.env['restaurant.floor'].create({
+                **model_data,
+                "pos_config_ids": [(4, self.id)],
+            })
+
+        def is_valid_floor_image(att):
+            return (
+                    att.exists()
+                    and att.res_model == 'restaurant.floor'
+                    and not att.res_field
+            )
+
+        IrAttachment = self.env['ir.attachment']
+        floor_image_ids = set(IrAttachment.search([
+            ('res_model', '=', 'restaurant.floor'),
+            ('res_id', '=', floor.id),
+        ]).ids)
+
+        # Create new attachment for newly uploaded images and link them to the floor
+        new_image_ids = floor_data.get("newImages", [])
+        attachments = IrAttachment.browse(new_image_ids)
+        for attachment in attachments:
+            if is_valid_floor_image(attachment) and not attachment.res_id:
+                attachment.res_id = floor.id
+                floor_image_ids.add(attachment.id)
+
+        # Decoration images
+        decoration_images = []
+        for decoration in layout_data.get('decorations', []):
+            if decoration.get('id') and decoration.get("shape") == "image":
+                decoration_images.append(decoration)
+
+        # Background image
+        bg_image = layout_data.get('bgImage')
+        if bg_image and bg_image.get('id'):
+            decoration_images.append(bg_image)
+
+        # If the same image is used in different floor we need to duplicate it and assign the correct res_id
+        for image in decoration_images:
+            image_id = image.get('id')
+            if image_id in floor_image_ids:
+                continue
+
+            attachment = IrAttachment.browse(image_id)
+            if is_valid_floor_image(attachment):
+                new_attachment = attachment.copy({"res_id": floor.id})
+                floor_image_ids.add(new_attachment.id)
+                image["id"] = new_attachment.id
+
+        floor.floor_plan_layout = layout_data
+        return floor
+
+    def _save_floor_plan_table(self, table_data, floor, existing_tables):
+        self.ensure_one()
+        table_id = table_data.get("id")
+        model_data, layout_data = self._split_layout_server_data(table_data, {"table_number", "seats"}, {"id"})
+        if table_id:
+            table = existing_tables.get(table_id)
+            if not table:
+                return None
+            table.write({**model_data, "floor_plan_layout": layout_data})
+        else:
+            table = self.env['restaurant.table'].create({
+                **model_data,
+                "floor_id": floor.id,
+                "floor_plan_layout": layout_data,
+            })
+
+        return table
+
+    @api.model
+    def _split_layout_server_data(self, data, model_keys, layout_exclude_keys):
+        layout_exclude = set(layout_exclude_keys) | set(model_keys)
+        model_data = {k: data[k] for k in model_keys if k in data}
+        layout_data = {k: v for k, v in data.items() if k not in layout_exclude}
+        return model_data, layout_data

--- a/addons/pos_restaurant/models/pos_restaurant.py
+++ b/addons/pos_restaurant/models/pos_restaurant.py
@@ -1,8 +1,24 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models, _, Command
+import uuid
+from base64 import b64decode
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models, _
 from odoo.exceptions import UserError
+from odoo.tools.mimetypes import guess_mimetype
+
+FLOOR_PLAN_SUPPORTED_IMAGE_MIMETYPES = {
+    'image/gif': '.gif',
+    'image/jpe': '.jpe',
+    'image/jpeg': '.jpeg',
+    'image/jpg': '.jpg',
+    'image/png': '.png',
+    'image/svg+xml': '.svg',
+    'image/webp': '.webp',
+}
 
 
 class RestaurantFloor(models.Model):
@@ -14,12 +30,10 @@ class RestaurantFloor(models.Model):
 
     name = fields.Char('Floor Name', required=True)
     pos_config_ids = fields.Many2many('pos.config', string='Point of Sales', domain="[('module_pos_restaurant', '=', True)]")
-    background_image = fields.Binary('Background Image')
-    background_color = fields.Char('Background Color', help='The background color of the floor in a html-compatible format')
     table_ids = fields.One2many('restaurant.table', 'floor_id', string='Tables')
     sequence = fields.Integer('Sequence', default=1)
     active = fields.Boolean(default=True)
-    floor_background_image = fields.Image(string='Floor Background Image')
+    floor_plan_layout = fields.Json(string='Floor Plan Layout')
 
     @api.model
     def _load_pos_data_domain(self, data, config):
@@ -27,7 +41,7 @@ class RestaurantFloor(models.Model):
 
     @api.model
     def _load_pos_data_fields(self, config):
-        return ['name', 'background_color', 'table_ids', 'sequence', 'pos_config_ids', 'floor_background_image']
+        return ['name', 'table_ids', 'sequence', 'pos_config_ids']
 
     @api.ondelete(at_uninstall=False)
     def _unlink_except_active_pos_session(self):
@@ -55,23 +69,6 @@ class RestaurantFloor(models.Model):
 
         return super().write(vals)
 
-    @api.model
-    def sync_from_ui(self, name, background_color, config_id):
-        floor_fields = {
-            "name": name,
-            "background_color": background_color,
-        }
-        pos_floor = self.create(floor_fields)
-        pos_floor.pos_config_ids = [Command.link(config_id)]
-        return {
-            'id': pos_floor.id,
-            'name': pos_floor.name,
-            'background_color': pos_floor.background_color,
-            'table_ids': [],
-            'sequence': pos_floor.sequence,
-            'tables': [],
-        }
-
     def deactivate_floor(self, session_id):
         draft_orders = self.env['pos.order'].search([('session_id', '=', session_id), ('state', '=', 'draft'), ('table_id.floor_id', '=', self.id)])
         if draft_orders:
@@ -82,6 +79,41 @@ class RestaurantFloor(models.Model):
 
         return True
 
+    @api.model
+    def add_floor_plan_image(self, name, data):
+        img = b64decode(data)
+        mimetype = guess_mimetype(img)
+        if mimetype not in FLOOR_PLAN_SUPPORTED_IMAGE_MIMETYPES:
+            raise UserError(_("Uploaded image's format is not supported. Try with: %s", ', '.join(FLOOR_PLAN_SUPPORTED_IMAGE_MIMETYPES.values())))
+        IrAttachment = self.env['ir.attachment']
+        checksum = IrAttachment._compute_checksum(img)
+        if not name:
+            name = f"{str(uuid.uuid4())[:6]}{FLOOR_PLAN_SUPPORTED_IMAGE_MIMETYPES[mimetype]}"
+        attachment_data = {
+            'name': name,
+            'res_model': 'restaurant.floor',
+            'res_id': 0,  # linked later
+        }
+        domain = [(field, '=', value) for field, value in attachment_data.items()]
+        domain.append(('checksum', '=', checksum))
+        attachment = IrAttachment.search(domain, limit=1) or None
+        if not attachment:
+            attachment_data['raw'] = img
+            attachment = IrAttachment.create(attachment_data)
+        return {'id': attachment.id, 'url': attachment.image_src}
+
+    @api.autovacuum
+    def _gc_embeddings(self):
+        """
+        Autovacuum: Cleanup floor plan images not associated
+        """
+        self.env['ir.attachment'].search([
+            ('res_model', '=', 'restaurant.floor'),
+            ('res_field', '=', False),
+            ('res_id', '=', 0),
+            ('write_date', '<', fields.Datetime.now() - relativedelta(days=1))
+        ]).unlink()
+
 
 class RestaurantTable(models.Model):
     _name = 'restaurant.table'
@@ -91,17 +123,11 @@ class RestaurantTable(models.Model):
 
     floor_id = fields.Many2one('restaurant.floor', string='Floor', index='btree_not_null')
     table_number = fields.Integer('Table Number', required=True, help='The number of the table as displayed on the floor plan', default=0)
-    shape = fields.Selection([('square', 'Square'), ('round', 'Round')], string='Shape', required=True, default='square')
-    position_h = fields.Float('Horizontal Position', default=10,
-        help="The table's horizontal position from the left side to the table's center, in pixels")
-    position_v = fields.Float('Vertical Position', default=10,
-        help="The table's vertical position from the top to the table's center, in pixels")
-    width = fields.Float('Width', default=50, help="The table's width in pixels")
-    height = fields.Float('Height', default=50, help="The table's height in pixels")
     seats = fields.Integer('Seats', default=1, help="The default number of customer served at this table.")
-    color = fields.Char('Color', help="The table's color, expressed as a valid 'background' CSS property value")
     parent_id = fields.Many2one('restaurant.table', string='Parent Table', help="The parent table if this table is part of a group of tables")
+    parent_side = fields.Selection(string='Parent Side', selection=[('left', 'Left'), ('right', 'Right'), ('top', 'Top'), ('bottom', 'Bottom')], help="The parent table side where this table will be displayed")
     active = fields.Boolean('Active', default=True, help='If false, the table is deactivated and will not be available in the point of sale')
+    floor_plan_layout = fields.Json(string='Floor Plan Layout')
 
     @api.depends('table_number', 'floor_id')
     def _compute_display_name(self):
@@ -114,7 +140,7 @@ class RestaurantTable(models.Model):
 
     @api.model
     def _load_pos_data_fields(self, config):
-        return ['table_number', 'width', 'height', 'position_h', 'position_v', 'parent_id', 'shape', 'floor_id', 'color', 'seats', 'active']
+        return ['table_number', 'parent_id', 'parent_side', 'floor_id', 'seats', 'active']
 
     def are_orders_still_in_draft(self):
         draft_orders_count = self.env['pos.order'].search_count([('table_id', 'in', self.ids), ('state', '=', 'draft')])
